File: Arrays/Medium/pb8NextPermutation.java
package Arrays.Medium;

public class pb8NextPermutation {
    public static void reverse(int arr[], int left, int right) {
        while (left < right) {
            int temp = arr[left];
            arr[left] = arr[right];
            arr[right] = temp;
            left++;
            right--;
        }
    }

    public static void swap(int arr[], int left, int right) {
        int temp = arr[left];
        arr[left] = arr[right];
        arr[right] = temp;
    }

    public static int[] nextPermutation(int arr[]) {
        int index = -1;
        for (int i = arr.length - 2; i > 0; i--) {
            if (arr[i] < arr[i + 1]) {
                index = i;
                break;
            }
        }
        if (index == -1) {
            reverse(arr, 0, arr.length - 1);
            return arr;
        } else {
            for (int i = arr.length - 1; i > index; i--) {
                if (arr[i] > arr[index]) {
                    swap(arr, i, index);
                    break;
                }
            }
            if (index + 1 < arr.length - 1) {
                reverse(arr, index + 1, arr.length - 1);
            }
        }
        return arr;
    }

    public static void main(String[] args) {
        int[] nums = { 2, 1, 5, 4, 3, 0, 0 };
        int nextPerm[] = nextPermutation(nums);
        System.out.print("[");
        for (Integer a : nextPerm) {
            System.out.print(a + ",");
        }
        System.out.print("]");
    }
}

File: Arrays/Medium/pb8_NextPermutation.js
// Helper function to swap two elements in an array.
function swap(nums, start, end) {
    [nums[start], nums[end]] = [nums[end], nums[start]];
}

// Helper function to reverse a portion of an array.
function reverse(nums, start, end) {
    while (start < end) {
        [nums[start], nums[end]] = [nums[end], nums[start]];
        start++;
        end--;
    }
}

var nextPermutation = function (nums) {
    // Find the first element from the right that is smaller than the element to its right (the pivot).
    let index = -1;

    for (let i = nums.length - 2; i >= 0; i--) {
        if (nums[i] < nums[i + 1]) {
            index = i;
            break;
        }
    }

    // If no pivot is found, the array is the largest permutation; reverse it to get the smallest permutation.
    if (index === -1) reverse(nums, 0, nums.length - 1);
    else {
        // Find the smallest element to the right of the pivot that is just larger than the pivot.
        for (let i = nums.length - 1; i > index; i--) {
            if (nums[i] > nums[index]) {
                // Swap the pivot with this larger element.
                swap(nums, index, i);
                break;
            }
        }
        // Reverse the subarray to the right of the pivot's original position to get the next lexicographical permutation.
        reverse(nums, index + 1, nums.length - 1);
    }
};
